File: algorithms/3_hard/S/00862_shortest_subarray_with_sum_at_least_k/solution.c
#define INF INT_MAX

struct pair {
    int idx;
    int64_t sum;
};

int shortestSubarray(const int *restrict nums, const int len, const int k) {
    if (k <= 0 || len <= 0 || nums == NULL) { return -1; }

    int res = INF;
    int64_t sum = 0;
    struct pair *restrict q = malloc(((size_t)len + 1) * sizeof(*q));
    if (q == NULL) { return -1; }
    q[0] = (struct pair){ .idx = -1, .sum = 0 };
    size_t q_head = 0;
    size_t q_len = 1;

    for (int i = 0; i < len; i += 1) {
        sum += nums[i];

        while (q_len > 0 && q[q_head + q_len - 1].sum >= sum) {
            q_len -= 1;
        }

        while (q_len > 0 && sum - q[q_head].sum >= k) {
            res = MIN(res, i - q[q_head].idx);
            q_len -= 1;
            q_head += 1;
        }

        q[q_head + q_len] = (struct pair){ .idx = i, .sum = sum};
        q_len += 1;
    }

    free(q);
    if (res == INF) {
        return -1;
    }
    return res;
}

File: algorithms/3_hard/S/00862_shortest_subarray_with_sum_at_least_k/solution.rs
use std::collections::VecDeque;

impl Solution {
    pub fn shortest_subarray(nums: Vec<i32>, k: i32) -> i32 {
        assert!(k > 0);
        let k = k as i64;
        let mut res: Option<isize> = None;
        let mut sum = 0_i64;
        let mut q: VecDeque<(isize, i64)> = VecDeque::new();
        q.push_back((-1, 0));

        for (i, num) in nums.into_iter().enumerate() {
            sum += num as i64;

            while let Some((_, val)) = q.back() && *val >= sum {
                q.pop_back();
            }

            while let Some((idx, val)) = q.front() && sum - *val >= k {
                res = Some(res.unwrap_or(isize::MAX).min((i as isize) - idx));
                q.pop_front();
            }

            q.push_back((i as isize, sum));
        }

        match res {
            Some(val) => val.try_into().unwrap(),
            None => -1,
        }
    }
}
